In [1]:
#Question 3a

In [2]:
import numpy as np
import itertools
import copy
import random

In [3]:
values = {'s1':8,'s2':11,'s3':9,'s4':12,'s5':14,'s6':10,'s7':6,'s8':7,'s9':13}
weights = {'s1':1,'s2':2,'s3':3,'s4':2,'s5':3,'s6':4,'s7':1,'s8':5,'s9':3}
limit = 16

In [4]:
n = 9
combinations = [list(i) for i in itertools.product([0, 1], repeat=n)]#all possible combinations

In [5]:
initial3 = [0,0,0,0,0,0,0,0,0]

In [6]:
def calc_weight(solution):
    weight = 0
    for count,bit in enumerate(solution):
        i = count+1
        weight = weight + solution[count]*weights["s"+str(i)]
    return weight

def calc_value(solution):
    value = 0
    for count,bit in enumerate(solution):
        i = count+1
        value = value + solution[count]*values["s"+str(i)]
    return value

In [7]:
def Neighbourhood(current_solution):
    possible_neighbours=[]
    for count,bit in enumerate(current_solution):
        new_neighbour = copy.copy(current_solution)
        if(bit==1):
            bit = 0
        else: bit = 1
        new_neighbour[count] = bit
        possible_neighbours.append(new_neighbour)
    return possible_neighbours

In [8]:
print(Neighbourhood(initial3))

[[1, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 0, 0, 0, 0, 0, 0], [0, 0, 1, 0, 0, 0, 0, 0, 0], [0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 1, 0, 0, 0, 0], [0, 0, 0, 0, 0, 1, 0, 0, 0], [0, 0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 0], [0, 0, 0, 0, 0, 0, 0, 0, 1]]


In [9]:
def random_walk(initial):
    walk_count = 0
    landscape = []
    current_solution = initial
    while(walk_count<100):
        walk_count+=1
        weight = calc_weight(initial)
        max_value = calc_value(current_solution)
        sol_list = []
        for neighbour in Neighbourhood(current_solution):
            value = calc_value(neighbour)
            weight = calc_weight(neighbour)
            attributes = {"vector":neighbour,"value":value,"weight":weight}
            sol_list.append(attributes)
        temp_max = max_value
        temp_solution = current_solution
        #print("\n")
        test = random.choice(sol_list)
        if(test['weight']>16):
            test = random.choice(sol_list)
        landscape.append(test['value'])
        current_solution = test['vector']
        
    return landscape    
    
    

In [10]:
walk = random_walk(initial3)

In [11]:
print(walk)

[10, 0, 8, 20, 34, 45, 51, 64, 56, 42, 36, 45, 55, 43, 55, 46, 34, 43, 33, 39, 46, 35, 26, 34, 27, 39, 26, 18, 31, 18, 31, 18, 28, 41, 50, 38, 46, 33, 45, 58, 48, 35, 27, 41, 49, 35, 49, 56, 50, 41, 47, 35, 29, 38, 30, 21, 29, 22, 29, 40, 26, 39, 31, 20, 31, 20, 30, 23, 13, 0, 12, 19, 33, 39, 32, 26, 34, 41, 33, 46, 39, 27, 13, 24, 34, 43, 49, 61, 69, 56, 44, 38, 30, 21, 29, 43, 55, 47, 54, 43]


In [12]:
var = np.var(walk) #variance
m=len(walk)
d = 1
m_d = m-d
r = ((1/(m_d)*var))
ave = sum(walk)/len(walk)
result = 0
for count,entry in enumerate(walk[:-d]):
    result += (entry-ave)*(walk[count]-ave)
r = r*result
print("r(1)=",r)

r(1)= 33025.10571874999


In [13]:
L = 1/(np.log(r))
#now to normalise
#Diam(G) = the max distance between [0,0,0,0,0,1,0,1,0] and [1,1,1,1,1,0,1,0,1] which corresponds
#to weight 15 (this is max moves without exceeding weight limit). This is 9 moves. Remove the two and add the rest.   
epsilon = L/9
print("xi:",epsilon)
    

xi: 0.010678602784559484


In [14]:
#Question 3b
U = []
for entry in Neighbourhood(initial3):
    for pos in Neighbourhood(entry):
        if(pos.count(1)==2):
            U.append(pos)
U = list(set(tuple(i)for i in U)) #removing any duplicates from process of making U
U = list(list(i)for i in U)
#for sanity sake here is the length of U
len(U)
    

36

In [15]:
def find_solution(initial):
    iteration = -1
    weight = calc_weight(initial)
    current_solution = initial
    max_value = calc_value(current_solution)
    print("Starting point = \t  ",initial,"value = ",max_value)
    while(weight<16):
        iteration+=1
        sol_list = []
        for neighbour in Neighbourhood(current_solution): #creates a dictionary with the neighbourhood and
                                                           #each of the potential solutions attributes
            value = calc_value(neighbour)
            weight = calc_weight(neighbour)
            attributes = {"vector":neighbour,"value":value,"weight":weight}
            sol_list.append(attributes)
        temp_max = max_value
        temp_solution = current_solution
        for entry in sol_list: #check the validty of each neighbourhood solution and decide if it is a new best
            if(entry['value']>temp_max and entry['weight']<=16):
                temp_solution = entry['vector']
                temp_max = entry['value']
        if(temp_max>max_value):
            current_solution = temp_solution
            max_value = temp_max
            
            
        elif(temp_max<=max_value):
            #print("\n No new solution found")
            break
    return current_solution,iteration,max_value

In [16]:
lengths = []
solutions = []
for initial in U:
    sol,it,value = find_solution(initial)
    print("took",it,"to find solution = ",sol,"value = ",value,"\n")
    lengths.append(it)
    solutions.append(sol)


Starting point = 	   [0, 1, 0, 0, 0, 0, 0, 1, 0] value =  18
took 4 to find solution =  [1, 1, 0, 1, 1, 0, 0, 1, 1] value =  65 

Starting point = 	   [0, 0, 0, 0, 0, 1, 0, 1, 0] value =  17
took 3 to find solution =  [1, 0, 0, 0, 1, 1, 0, 1, 1] value =  52 

Starting point = 	   [1, 0, 0, 0, 1, 0, 0, 0, 0] value =  22
took 5 to find solution =  [1, 1, 0, 1, 1, 1, 1, 0, 1] value =  74 

Starting point = 	   [0, 0, 1, 0, 1, 0, 0, 0, 0] value =  23
took 5 to find solution =  [1, 1, 1, 1, 1, 0, 1, 0, 1] value =  73 

Starting point = 	   [1, 0, 1, 0, 0, 0, 0, 0, 0] value =  17
took 5 to find solution =  [1, 1, 1, 1, 1, 0, 1, 0, 1] value =  73 

Starting point = 	   [1, 0, 0, 0, 0, 0, 0, 1, 0] value =  15
took 4 to find solution =  [1, 1, 0, 1, 1, 0, 0, 1, 1] value =  65 

Starting point = 	   [0, 1, 0, 0, 0, 0, 1, 0, 0] value =  17
took 5 to find solution =  [1, 1, 0, 1, 1, 1, 1, 0, 1] value =  74 

Starting point = 	   [0, 0, 0, 0, 0, 1, 0, 0, 1] value =  23
took 5 to find solution =  [1

In [17]:
print("LLM(U)=",(sum(lengths)/len(lengths)))

LLM(U)= 4.694444444444445


In [18]:
#Question 3c
#Assuming the mathematical definition of a set as mentioned in the question where all elements must be unique.
#making sure there are no duplicates in the set O.
O = list(set(tuple(i)for i in solutions)) #removing any duplicates from process of making O
O = list(list(i)for i in O)
#for sanity sake here is the length of O
print(len(O))
O

7


[[1, 1, 1, 1, 1, 0, 1, 0, 1],
 [1, 1, 0, 1, 1, 0, 0, 1, 1],
 [0, 0, 1, 1, 1, 0, 0, 1, 1],
 [1, 0, 1, 1, 1, 1, 0, 0, 1],
 [1, 0, 0, 0, 1, 1, 0, 1, 1],
 [1, 1, 0, 1, 1, 1, 1, 0, 1],
 [0, 1, 0, 1, 1, 0, 1, 1, 1]]

In [19]:
def Amp(P):
    vals = [] #array containing all fitness values (value of knapsack for given solution set)
    for element in P:
        vals.append(calc_value(element))
    
    return (len(P)*(max(vals)-min(vals)))/sum(vals) #formula for Amp(P)

In [20]:
Amp(U)

0.7

In [21]:
Amp(O)

0.34375

In [22]:
delta_amp = (Amp(U)-Amp(O))/Amp(U)
delta_amp

0.5089285714285714

In [23]:
def Gap(P):
    vals = []
    for element in P:
        vals.append(calc_value(element))
    array = []
    for val in vals:
        array.append(val-max(vals))
    return sum(array)/(len(P)*max(vals))
    

In [24]:
Gap(O)

-0.13513513513513514

With a small gap it is a relatively easy problem to solve, LLM =4.7 means that there is a short distance to find the correct
solution. With such a low value for xi(normalised correlation length) the landscape is not flat or smooth and as such there are many local optima to get trapped in. Combined with a delta amp being not very high and therefore not a large difference between the best solution and the worst it can be concluded that while this is not a difficult problem, it is not the easiest. It would not be very hard to solve the problem.